### ISSUE TITLE
[YSQL] Incorrect Results on TAQO complex query with distinct pushdown

### ISSUE THREAD

**Issue by timothy-e** [yugabyte/yugabyte-db#22615]
Jira Link: [DB-11524](https://yugabyte.atlassian.net/browse/DB-11524)
### Description

The same query, but using distinct index pushdown on a different index, produces different results:

```sql
yugabyte# EXPLAIN ANALYZE
SELECT DISTINCT c_int FROM t50000 WHERE c_int IN (58, 19) ORDER BY c_int DESC;
                                                                 QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------
 Unique  (cost=0.00..66.50 rows=500 width=4) (actual time=1.219..1.222 rows=2 loops=1)
   ->  Index Only Scan Backward using t50000_1_idx on t50000  (cost=0.00..65.25 rows=500 width=4) (actual time=1.217..1.219 rows=2 loops=1)
         Index Cond: (c_int = ANY ('{58,19}'::integer[]))
         Heap Fetches: 0
 Planning Time: 0.272 ms
 Execution Time: 1.270 ms
 Peak Memory Usage: 24 kB
(7 rows)

yugabyte# EXPLAIN ANALYZE /*+ IndexOnlyScan(t50000) */
SELECT DISTINCT c_int FROM t50000 WHERE c_int IN (58, 19) ORDER BY c_int DESC;
                                                                     QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Unique  (cost=0.00..61.65 rows=500 width=4) (actual time=1.225..1.226 rows=1 loops=1)
   ->  Distinct Index Only Scan Backward using t50000_4_idx on t50000  (cost=0.00..61.65 rows=500 width=4) (actual time=1.222..1.224 rows=1 loops=1)
         Index Cond: (c_int = ANY ('{58,19}'::integer[]))
         Distinct Keys: c_int
         Heap Fetches: 0
 Planning Time: 0.312 ms
 Execution Time: 1.287 ms
 Peak Memory Usage: 24 kB
(8 rows)
```

I'm not sure why the hint causes it to use `t50000_4_idx`, but the results are incorrect. 

schema:
```sql
CREATE TABLE t50000
AS
SELECT c_int,
       (case when c_int % 2 = 0 then true else false end) as c_bool,
       (c_int + 0.0001)::text as c_text,
        (c_int + 0.0002):: varchar as c_varchar,
        (c_int + 0.1):: decimal as c_decimal,
        (c_int + 0.2):: float as c_float,
        (c_int + 0.3):: real as c_real,
        (c_int + 0.4) ::money as c_money FROM generate_Series (1, 5000) c_int;
CREATE INDEX t50000_1_idx ON t50000 (c_int ASC);
CREATE INDEX t50000_2_idx ON t50000 (c_int ASC, c_bool ASC);
CREATE INDEX t50000_3_idx ON t50000 (c_int ASC, c_text ASC);
CREATE INDEX t50000_4_idx ON t50000 (c_int ASC, c_varchar ASC);
CREATE INDEX t50000_5_idx ON t50000 (c_float ASC, c_text ASC, c_varchar ASC);
CREATE INDEX t50000_6_idx ON t50000 (c_float ASC, c_decimal ASC, c_varchar ASC);
CREATE INDEX t50000_7_idx ON t50000 (c_float ASC, c_real ASC, c_money ASC);
```

### Issue Type

kind/bug

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-11524]: https://yugabyte.atlassian.net/browse/DB-11524?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pao214** [comment]
Tserver Log

```
I0530 15:09:51.020293 1810935808 scan_choices.cc:290] vlog2: Result<bool> yb::docdb::HybridScanChoices::SkipTargetsUpTo(Slice) Updating current target to be >= DocKey([], [58, 58])
I0530 15:09:51.020378 1810935808 scan_choices.cc:541] vlog2: After Result<bool> yb::docdb::HybridScanChoices::SkipTargetsUpTo(Slice) current_scan_target_ is Corruption (yb/dockv/doc_key.cc:53): Error when decoding range components of a document key: Unexpected end of key when decoding document key: 488000003A
I0530 15:09:51.020438 1810935808 scan_choices.cc:576] vlog2: IncrementScanTargetAtOptionList: Incrementing at 0
I0530 15:09:51.020464 1810935808 scan_choices.cc:688] vlog2: IncrementScanTargetAtOptionList: Key after increment is Corruption (yb/dockv/doc_key.cc:53): Error when decoding range components of a document key: Unexpected end of key when decoding document key: 4880000013
I0530 15:09:51.020488 1810935808 scan_choices.cc:745] vlog2: DoneWithCurrentTarget: Current_scan_target_ is DocKey([], [19]), result: 1
I0530 15:09:51.020504 1810935808 scan_choices.cc:775] vlog2: SeekToCurrentTarget: pos: 488000003A488000003A214A80 (SubDocKey(DocKey([], [58, 58]), [SystemColumnId(0)]))
```

**Comment by pao214** [comment]
Root cause:

1. We sometimes add a groupend marker to the current scan target (and sometimes not?). Perhaps, @tanujnay112 knows more about this logic of when we add a group end character to the current scan target. Code line: https://github.com/yugabyte/yugabyte-db/blob/c7105eadbcd599903e6f10f1a6b4a2fe468ef776/src/yb/docdb/scan_choices.cc#L737
2. When we add a groupend character to the scan target and we want to append kHighest (infinity), we need to do it before the groupend character or the kHighest (infinity) will not be in effect. Code line: https://github.com/yugabyte/yugabyte-db/blob/c7105eadbcd599903e6f10f1a6b4a2fe468ef776/src/yb/docdb/scan_choices.cc#L788

The fix involves appending the kHighest before the kGroupEnd character.

**Comment by pao214** [comment]
The fix is landed on master. I will wait for a bit before backporting to have some bake time.
cc @tanujnay112

**Comment by pao214** [comment]
Landed the fix on 2.20 and 2024.1 as well. Distinct pushdown support goes back until 2.20.
